Storing Generator Filter Information in NanoAOD

mzarucki opened the pull request:
#### PR description:
  - new nanoAOD module `SimpleFlatTableProducerBaseLumi` (`edm::one::EDProducer`) that stores the generator filter information in a flat table
  - the information is taken from the [GenFilterInfo](https://github.com/cms-sw/cmssw/blob/master/SimDataFormats/GeneratorProducts/interface/GenFilterInfo.h) header in the `LuminosityBlocks` tree in miniAOD
  - the module is configured to store the following generator filter information: `numEventsTotal`, `numEventsPassed`, `filterEfficiency`,`filterEfficiencyError`
  - the `genFilterTable` is added to the `globalTablesMC` sequence in [globals_cff.py](https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/python/globals_cff.py)
  - the `nanoaod::FlatTable` is written into the `LuminosityBlocks` tree as new branches via the added `LumiOutputBranches` type
  - the motivation (xsec/luminosity normalisation) and implementation was presented at the X-POG meeting (16.06.2021): [Storing Generator Filter Information in NanoAOD](https://indico.cern.ch/event/1039035/contributions/4406005/attachments/2265487/3846473/NanoAOD_GenFilterInfo_XPOG_16.06.2021_Zarucki.pdf)

#### PR validation:
The code has been tested in _CMSSW_12_0_0_pre1_, running a standard nanoAOD workflow on a miniAOD signal MC sample with a generator filter (_SMS-T2tt_dM-10to80_genHT-160_genMET-80_mWMin-0p1_), over several thousand events (several LS). The output root file correctly has the requested generator filter information stored.

Following rebasing on the latest master branch, the code has passes the nanoAOD unit tests that run over data (`runTheMatrix.py -l 136.8523`) in the latest IB _CMSSW_12_0_X_2021-06-17-1100_. The MC equivalent (`runTheMatrix.py -l 1325.81`) and `scram b runtests` fail due to a DAS error not being able to access the file. I believe it would still important to inspect the output of the nanoAOD MC unit tests.

mzarucki committed:
Add nanoAOD flat table producers to store generator filter information in lumi tree

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34169/23376

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/NanoAODOutputModule.cc  modified in PR(s): #32167

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34169/23376/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34169/23376/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34169/23377

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/NanoAODOutputModule.cc  modified in PR(s): #32167

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34169/23377/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34169/23377/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

mzarucki commented:
As discussed in yesterday's [X-POG meeting](https://indico.cern.ch/event/1039035/contributions/4406005/attachments/2265487/3846473/NanoAOD_GenFilterInfo_XPOG_16.06.2021_Zarucki.pdf), there might still be some open questions/comments:

- Using unsigned ints (`uint8`) in the [configuration](https://github.com/cms-sw/cmssw/pull/34169/files#diff-5b83d7a19d8fb45f37fa1bd2db0aeb66a38043fc710420e0f3cdf5675c24eb27R49-R50) for the `numEventsPassed` and `numEventsTotal` (as in the [GenFilterInfo](https://github.com/cms-sw/cmssw/blob/master/SimDataFormats/GeneratorProducts/interface/GenFilterInfo.h) header) gave nonsensical answers (passed > total), so a normal `int` was used
  - Is a precision of 6 for the `ints` sufficient?
- In order to avoid conflict with https://github.com/cms-sw/cmssw/commit/b0c0009e28b163aa678395192f3e5323c7984fd6, `std::vector<LumiOutputBranches> m_lumiTables2` had to be added that uses `nanoaod::FlatTables`
- Is there a simple way to make the code more efficient in terms of re-using existing modules? Some blocks of code were duplicated and modified to handle LuminosityBlocks and write to the Lumi tree


I leave the above considerations to the experts.


Cheers,
Mateusz

makortel commented on PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h:
```diff
@@ -271,3 +272,134 @@ class FirstObjectSimpleFlatTableProducer : public SimpleFlatTableProducerBase<T,
 };
+
+template <typename T, typename TProd>
+class SimpleFlatTableProducerBaseLumi : public edm::one::EDProducer<edm::EndLuminosityBlockProducer> {
```
Why does this module (template) need to be `edm::one`? Could it be `edm::stream`?

If it needs to stay `edm::one` for some reason, please use also [`edm::LuminosityBlockCache`](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkGlobalModuleInterface#edm_LuminosityBlockCacheT) extension to tell the framework the the module can process events from multiple LuminosityBlocks concurrently.

mzarucki replied:
Dear @makortel,

At the time of development I tried to use the `edm::stream` (as in _SimpleFlatTableProducerBase_) guided by the [_FWMultithreadedFrameworkStreamModuleInterface_ Twiki](https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkStreamModuleInterface#edm_BeginLuminosityBlockProducer) however, this resulted in a series of compilation errors that I was not able to resolve. Using `edm::one` solved the issues.

If you have any suggestions on how to make `edm::stream` work, I would be grateful.

Best,
Mateusz

makortel replied:
What kind of compilation errors did you get?

mzarucki replied:
The first one is: `/afs/cern.ch/work/m/mzarucki/nanoAOD/CMSSW_12_0_0_pre3/src/PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h:315:8: error: 'void SimpleFlatTableProducerBaseLumi<T, TProd>::endLuminosityBlockProduce(edm::LuminosityBlock&, const edm::EventSetup&) [with T = GenFilterInfo; TProd = GenFilterInfo]' marked 'final', but is not virtual`

which can be solved by just removing the `final` specifier for the `endLuminosityBlockProduce` method. 

A following compilation error is:

`/cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/src/FWCore/Framework/interface/stream/callAbilities.h:455:43: error: 'globalEndLuminosityBlockProduce' is not a member of 'LumiSingletonSimpleFlatTableProducer<GenFilterInfo>'
  455 |         T::globalEndLuminosityBlockProduce(Lumi, iES, iRC);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~`

makortel replied:
My bad, after further thought stream module type actually wouldn't help. I was thinking we could use the "streams" to work around the thread-unsafety of `StringObjectFunction` and `StringCutObjectSelector`, but the LuminosityBlock transitions of stream modules are global (i.e. can occur concurrently).

So the best way to support "processing Events from multiple LuminosityBlocks concurrently" would be to use the ` edm::LuminosityBlockCache` extension. Since you don't really need the cache, you could declare it as e.g. `edm::LuminosityBlockCache<int>` and in `globalBeginLuminosityBlock()` return `nullptr`.

mzarucki replied:
Dear @makortel,

Therefore, is `edm:one` fine to use for this module?

As suggested, I have added the `edm:LuminosityBlockCache` extension as follows: 
```
class SimpleFlatTableProducerBaseLumi : public edm::one::EDProducer<edm::EndLuminosityBlockProducer, edm::LuminosityBlockCache<int>> {
```

I have also added the `globalBeginLuminosityBlock` method which returns `nullptr`.

Unfortunately, I get the following compilation errors:
```
 from /afs/cern.ch/work/m/mzarucki/nanoAOD/CMSSW_12_0_0_pre3/src/PhysicsTools/NanoAOD/plugins/SimpleFlatTableProducerPlugins.cc:1:
/cvmfs/cms.cern.ch/slc7_amd64_gcc900/external/gcc/9.3.0/include/c++/9.3.0/bits/unique_ptr.h: In instantiation of 'typename std::_MakeUniq<_Tp>::__single_object std::make_unique(_Args&& ...) [with _Tp = LumiSingletonSimpleFlatTableProducer<GenFilterInfo>; _Args = {const edm::ParameterSet&}; typename std::_MakeUniq<_Tp>::__single_object = std::unique_ptr<LumiSingletonSimpleFlatTableProducer<GenFilterInfo>, std::default_delete<LumiSingletonSimpleFlatTableProducer<GenFilterInfo> > >]':
/cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/src/FWCore/Framework/src/MakeModuleHelper.h:39:40:   required from 'static std::unique_ptr<_Tp> edm::MakeModuleHelper<Base>::makeModule(const edm::ParameterSet&) [with T = LumiSingletonSimpleFlatTableProducer<GenFilterInfo>; Base = edm::one::EDProducerBase]'
/cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/src/FWCore/Framework/src/WorkerMaker.h:83:107:   required from 'std::shared_ptr<edm::maker::ModuleHolder> edm::WorkerMaker<T>::makeModule(const edm::ParameterSet&) const [with T = LumiSingletonSimpleFlatTableProducer<GenFilterInfo>]'
/cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_0_pre3/src/FWCore/Framework/src/WorkerMaker.h:77:40:   required from here
/cvmfs/cms.cern.ch/slc7_amd64_gcc900/external/gcc/9.3.0/include/c++/9.3.0/bits/unique_ptr.h:857:30: error: invalid new-expression of abstract class type 'LumiSingletonSimpleFlatTableProducer<GenFilterInfo>'
  857 |     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Is there something obvious that I might be missing here?

Best regards,
Mateusz

makortel replied:
The use of `LuminosityBlockCache` requires overriding also `void SimpleFlatTableProducerBaseLumi::globalEndLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) override {}`.
(sorry for not mentioning it earlier)
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkOneModuleInterface#edm_LuminosityBlockCacheT

mzarucki replied:
Thanks for the clarification!

I have made the discussed updates to support processing events from multiple LuminosityBlocks concurrently in https://github.com/cms-sw/cmssw/pull/34169/commits/a4a7f44db5a06f68f25b308f0d0e1b0e8534bb06 

I would be grateful if you could have a look whether this is what you were expecting.

Cheers,
Mateusz

makortel replied:
> I would be grateful if you could have a look whether this is what you were expecting.

Yes, it looks good now. Thanks!

makortel commented on PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h:
```diff
@@ -271,3 +272,134 @@ class FirstObjectSimpleFlatTableProducer : public SimpleFlatTableProducerBase<T,
+        extension_(params.existsAs<bool>("extension") ? params.getParameter<bool>("extension") : false),
+        skipNonExistingSrc_(
+            params.existsAs<bool>("skipNonExistingSrc") ? params.getParameter<bool>("skipNonExistingSrc") : false),
+        src_(skipNonExistingSrc_ ? mayConsume<TProd>(params.getParameter<edm::InputTag>("src"))
```
Why `mayConsume`?

mzarucki replied:
Dear @makortel,

Thank you very much for your comments.

I based the 'Lumi version' of the module off the existing _SimpleFlatTableProducerBase_: https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h#L23

Therefore, I think this question is best addressed to the original developers. If it were to be updated, then I think it would make sense that it is updated consistently in both modules by the original developers.

Best,
Mateusz

makortel commented on PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h:
```diff
@@ -271,3 +272,134 @@ class FirstObjectSimpleFlatTableProducer : public SimpleFlatTableProducerBase<T,
+  class Variable : public VariableBase {
+  public:
+    Variable(const std::string &aname, const edm::ParameterSet &cfg) : VariableBase(aname, cfg) {}
+    virtual void fill(std::vector<const T *> selobjs, nanoaod::FlatTable &out) const = 0;
```
Can the copy of the vector be avoided?
```suggestion
    virtual void fill(const std::vector<const T *>& selobjs, nanoaod::FlatTable &out) const = 0;
```

mzarucki replied:
Dear @makortel,

I based the 'Lumi version' of the module off the existing _SimpleFlatTableProducerBase_: https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h#L85

Therefore, I think this question is best addressed to the original developers. If it were to be updated, then I think it would make sense that it is updated consistently in both modules by the original developers.

Best,
Mateusz

makortel commented on PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h:
```diff
@@ -271,3 +272,134 @@ class FirstObjectSimpleFlatTableProducer : public SimpleFlatTableProducerBase<T,
+  const bool skipNonExistingSrc_;
+  const edm::EDGetTokenT<TProd> src_;
+
+  class VariableBase {
```
This class appears to be independent of `T` or `TProd`. Moving its definition outside of the `SimpleFlatTableProducerBaseLumi` would generate less code (only one definition instead of one per instantiation of `SimpleFlatTableProducerBaseLumi`).

mzarucki replied:
Dear @makortel,

As with the previous cases, I based the 'Lumi version' of the module off the existing SimpleFlatTableProducerBase: https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h#L65

Therefore, I think this question is best addressed to the original developers. If it were to be updated, then I think it would make sense that it is updated consistently in both modules by the original developers.

Best,
Mateusz

mariadalfonso commented:
in addition to makortel comments please run the following
>   You can also run `scram build code-format` to apply code format directly

mzarucki committed:
Apply code format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34169/23395

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/NanoAODOutputModule.cc  modified in PR(s): #32167

cmsbuild commented:
A new Pull Request was created by  @mzarucki (Mateusz Zarucki) for master.

It involves the following packages:

PhysicsTools/NanoAOD


@cmsbuild, @mariadalfonso, @gouskos, @fgolf can you please review it and eventually sign? Thanks.
@gpetruc, @swertz this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mariadalfonso commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-84e32b/16108/summary.html
**COMMIT**: 68d7a154f32b58aa9df577d85abf35ab08f38a5f
**CMSSW**: CMSSW_12_0_X_2021-06-18-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/34169/16108/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 38
 - DQMHistoTests: Total histograms compared: 2785631
 - DQMHistoTests: Total failures: 12
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2785596
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 37 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 160 log files, 37 edm output root files, 38 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-06-18-1100+84e32b/43560/triggerResults): found differences in 7 / 37 workflows

mariadalfonso commented:
@mzarucki 
can you please address the comments received ?
were you able to use the unsigned int ?

mzarucki commented:
Dear @makortel, @mariadalfonso,

> can you please address the comments received ?

Thank you for your comments. I have addressed each one separately, even though the majority are related to the original [_SimpleFlatTableProducerBase_](https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h#L15) that I based off the 'Lumi version' of the module. As already pointed out, I think those questions are best addressed to the original developers. If the updates are to be made, then I think it would make sense that it is updated consistently in both modules by the original developers.



> were you able to use the unsigned int ?

Concerning the unsigned ints, I already commented [above](https://github.com/cms-sw/cmssw/pull/34169#issuecomment-863433281) that using unsigned ints (`uint8`) in the configuration for the `numEventsPassed` and `numEventsTotal` gave nonsensical answers (passed > total), so a normal `int` was used.



In the meantime, I have tried to have a set up a proper testing setup, as after rebasing on master, my setup on `CMSSW_12_0_0_pre1` does not work any more. 

I managed to get the standard NanoAOD workflow on MC to work via adding the  `--ibeos` option, resolving the DAS error with accessing the file: `runTheMatrix.py -l 1325.81 --ibeos` (running over data does not make much sense in this case). Looking at the output, I see that the latest changes on master have changed the branch structure significantly:

`OBJ: TBranchElement	nanoaodFlatTable_genFilterTable__DQM.	nanoaodFlatTable_genFilterTable__DQM.`

instead of the four TBranches I saw earlier: `GenFilter_filterEfficiency`, `GenFilter_filterEfficiencyError`, `GenFilter_numEventsPassed`, `GenFilter_numEventsTotal`. I would be grateful if you could explain the implemented changes in the core code and whether is expected.

As described above, for a final validation, I was always running a standard nanoAOD workflow over a miniAOD signal MC sample with a generator filter (_SMS-T2tt_dM-10to80_genHT-160_genMET-80_mWMin-0p1_) to ensure that the generator filter efficiencies are correctly saved. Now the input sample is not valid any more (Error: `invalid value 3.40282e+38 for reco::GsfElectron.conversionRejection_.vtxFitProb`). 

Therefore, could you please point me to the latest nanoAOD recipe together with a valid MC sample that contains a generator filter? 

Thank you very much.

Best regards,
Mateusz

silviodonato commented:
enable profiling

silviodonato commented:
please test
(this is just to test a new feature of the profiling test)

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-84e32b/16796/summary.html
**COMMIT**: 68d7a154f32b58aa9df577d85abf35ab08f38a5f
**CMSSW**: CMSSW_12_0_X_2021-07-13-2300/slc7_amd64_gcc900
**Additional Tests**: PROFILING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/34169/16796/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 38
 - DQMHistoTests: Total histograms compared: 2786302
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2786279
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 37 files compared)
 - Checked 160 log files, 37 edm output root files, 38 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-07-13-2300+84e32b/44079/triggerResults): found differences in 7 / 37 workflows

mzarucki commented:
Dear NanoAOD experts,

Do you have answers to the [above](https://github.com/cms-sw/cmssw/pull/34169#issuecomment-866893899) questions? These are necessary to complete the validation of the changes in this PR.

Apart from these, it would be important to know how to proceed with the remainder of @makortel's comments, which are related to the original module that this one was based on.

Best regards,
Mateusz

makortel commented:
Given that an existing pattern was followed, I'd be fine if the other (than `LuminosityBlockCache`) comments are followed up later for all use cases (it's anyway up to `xpog`).

Especially for the `mayConcumes()` I'm suspecting the intended behavior to differ from the actual behavior.

mzarucki committed:
Update nanoAOD lumi flat table producers to support processing concurrent LS

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34169/24160

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/NanoAODOutputModule.cc  modified in PR(s): #32167
  - File  PhysicsTools/NanoAOD/python/globals_cff.py  modified in PR(s): #34329

cmsbuild commented:
Pull request #34169 was updated. @cmsbuild, @mariadalfonso, @gouskos, @fgolf can you please check and sign again.

mariadalfonso commented:
@mzarucki 
the PR has conflicts, please note that we moved from Sequences to Tasks

mariadalfonso commented:
> @mzarucki
> the PR has conflicts, please note that we moved from Sequences to Tasks
> 
> Question: did you checked that the PhysicsTools/NanoAOD/test/inspectNanoFile.py correctly document the addition in the lumiTree ? If not please update it

the script in updated in https://github.com/cms-sw/cmssw/pull/34985

mariadalfonso commented:
@mzarucki can you rebase to the most recent IB

mariadalfonso commented:
@kdlong  do we need to add this in the nanoGEN ?
https://github.com/cms-sw/cmssw/blob/f904c8155ac865af2a57381cfda7ccfa9ea4337c/PhysicsTools/NanoAOD/python/nanogen_cff.py#L28

mariadalfonso commented:
@mzarucki please resolve conflicts
@kdlong see comment in https://github.com/cms-sw/cmssw/pull/34169#issuecomment-919778222

mzarucki committed:
Merge branch 'master' into nanoAOD_genFilterEff_120X

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34169/27145

- This PR adds an extra 32KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h  modified in PR(s): #36037
  - File  PhysicsTools/NanoAOD/plugins/NanoAODOutputModule.cc  modified in PR(s): #32167
  - File  PhysicsTools/NanoAOD/plugins/SimpleFlatTableProducerPlugins.cc  modified in PR(s): #36037
  - File  PhysicsTools/NanoAOD/python/globals_cff.py  modified in PR(s): #36037

cmsbuild commented:
Pull request #34169 was updated. @cmsbuild, @mariadalfonso, @gouskos, @fgolf can you please check and sign again.

mzarucki commented:
Dear @mariadalfonso,

With the main data-taking periods of this year being over, I have now had the time to revisit this PR.

> @mzarucki the PR has conflicts, please note that we moved from Sequences to Tasks

I have fixed the merge conflict by updating from Sequences to Tasks: https://github.com/cms-sw/cmssw/pull/34169/commits/7ea752f38652cc716bbae6b8f6bfecc2055cd4e8

> Question: did you checked that the PhysicsTools/NanoAOD/test/inspectNanoFile.py correctly document the addition in the lumiTree ? If not please update it
> the script in updated in #34985

I am not sure what requires updating in that script. Could you please clarify?

I have still not received a response on the [above](https://github.com/cms-sw/cmssw/pull/34169#issuecomment-866893899) question concerning the branch structure:

> Looking at the output, I see that the latest changes on master have changed the branch structure significantly:
> OBJ: TBranchElement nanoaodFlatTable_genFilterTable__DQM. nanoaodFlatTable_genFilterTable__DQM.
> instead of the four TBranches I saw earlier: GenFilter_filterEfficiency, GenFilter_filterEfficiencyError, GenFilter_numEventsPassed, GenFilter_numEventsTotal. I would be grateful if you could explain the implemented changes in the core code and whether is expected.

The current branch structure of the FlatTable is rather different to when the code was first developed, so it is not clear to me how to properly re-validate the updates. Could you please clarify the changes and point me to documentation on how to properly access the values of the TBranchElements?

Best,
Mateusz

mariadalfonso commented:
please test

mariadalfonso commented:
https://cms-nanoaod-integration.web.cern.ch/integration/34169/mc106Xul17v2_size_report.html

For new branches are added in the lumi-Block
<img width="620" alt="Screen Shot 2021-12-07 at 16 56 51" src="https://user-images.githubusercontent.com/5656714/145062895-1ca8a9a8-091c-47d7-a40e-0b232dceaa97.png">

[ This is the result of the inspection of the PhysicsTools/NanoAOD/test/inspectNanoFile.py ]

mariadalfonso commented:
> > Question: did you checked that the PhysicsTools/NanoAOD/test/inspectNanoFile.py correctly document the addition in the lumiTree ? If not please update it
> > the script in updated in #34985
> 
> I am not sure what requires updating in that script. Could you please clarify?

nothing to do from your side.

> I have still not received a response on the [above](https://github.com/cms-sw/cmssw/pull/34169#issuecomment-866893899) question concerning the branch structure:
> 
> > Looking at the output, I see that the latest changes on master have changed the branch structure significantly:
> > OBJ: TBranchElement nanoaodFlatTable_genFilterTable__DQM. nanoaodFlatTable_genFilterTable__DQM.
> > instead of the four TBranches I saw earlier: GenFilter_filterEfficiency, GenFilter_filterEfficiencyError, GenFilter_numEventsPassed, GenFilter_numEventsTotal. I would be grateful if you could explain the implemented changes in the core code and whether is expected.
> 
> The current branch structure of the FlatTable is rather different to when the code was first developed, so it is not clear to me how to properly re-validate the updates. Could you please clarify the changes and point me to documentation on how to properly access the values of the TBranchElements?
> 

nothing really changed:
you are looking at the nano in the EDM + DQM files, you can also check the flat tree directly.
the cms-driver command you used now and have theses 

    --eventcontent NANOAODSIM \
    --datatier NANOAODSIM \
    --step NANO \

mariadalfonso commented:
The only things left form my side is this request to add to the nanoGEN sequence
https://github.com/cms-sw/cmssw/pull/34169#issuecomment-919778222
please check that the workflow 546,547,548 runs

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-84e32b/21046/summary.html
**COMMIT**: 7ea752f38652cc716bbae6b8f6bfecc2055cd4e8
**CMSSW**: CMSSW_12_3_X_2021-12-06-2300/slc7_amd64_gcc900
**Additional Tests**: PROFILING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/34169/21046/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 42
 - DQMHistoTests: Total histograms compared: 3250608
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3250586
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 41 files compared)
 - Checked 177 log files, 37 edm output root files, 42 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2021-12-06-2300+84e32b/47295/triggerResults): no differences found

mzarucki commented:
Dear @mariadalfonso,

Thank you for clarifying and checking!

I did not realise that `1325.81` uses the `NANOEDMAODSIM` content (and not `NANOAODSIM`). I also managed to run the `inspectNanoFile.py` on the output, which (as you already showed) correctly documents the addition to the lumiTree.

Concerning nanoGEN, I understand you are waiting for @kdlong's feedback.

The only other thing that comes to mind is the `int` vs `uint` usage for the `numEventsPassed` and `numEventsTotal` (as discussed [above](https://github.com/cms-sw/cmssw/pull/34169#issuecomment-863433281)) - I was not able to debug this further, so it is left as a standard `int`. I would guess that the issues are beyond this particular module.


Best regards,
Mateusz

kdlong commented:
Sorry I didn't look closely at the technical implementation here, but it would be very useful to have this in NanoGen.

Note also if we simply stored the cross section of the sample in the Nano, with the gen filter efficiency, you could run the GenXSecAnalyzer on NanoAOD which would be very convenient.

mzarucki commented:
Hi,

If you point me where exactly this should be added, I will be happy to update it in this PR.

Cheers,
Mateusz

kdlong commented:
@mzarucki: Either just add genFilterTable here (with genTable): https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/NanoAOD/python/nanogen_cff.py#L28

or define a new task excludes the puTable and replace genTable with that. Thanks!

mzarucki committed:
Add genFilterTable to the NanoGen sequence

mzarucki commented:
Dear @kdlong, @mariadalfonso,

I have added the `genFilterTable` into the NanoGen sequence in this commit: https://github.com/cms-sw/cmssw/pull/34169/commits/27fcb91fecfc2f4b6c0b56d680f1bda33a61ee76

I ran the workflows: `runTheMatrix.py -l 546,547,548 --ibeos` with no failed tests.

Best regards,
Mateusz

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-34169/27267

- This PR adds an extra 12KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/interface/SimpleFlatTableProducer.h  modified in PR(s): #36037
  - File  PhysicsTools/NanoAOD/plugins/NanoAODOutputModule.cc  modified in PR(s): #32167
  - File  PhysicsTools/NanoAOD/plugins/SimpleFlatTableProducerPlugins.cc  modified in PR(s): #36037
  - File  PhysicsTools/NanoAOD/python/globals_cff.py  modified in PR(s): #36037
  - File  PhysicsTools/NanoAOD/python/nanogen_cff.py  modified in PR(s): #32167

cmsbuild commented:
Pull request #34169 was updated. @cmsbuild, @mariadalfonso, @gouskos, @fgolf can you please check and sign again.

mariadalfonso commented:
please test workflow 546,547,548

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-84e32b/21129/summary.html
**COMMIT**: 27fcb91fecfc2f4b6c0b56d680f1bda33a61ee76
**CMSSW**: CMSSW_12_3_X_2021-12-09-1100/slc7_amd64_gcc900
**Additional Tests**: PROFILING
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/34169/21129/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-84e32b/546.0_DYToLL_M-50_13TeV_pythia8+DYToLL_M-50_13TeV_pythia8+NANOGENFromGen
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-84e32b/547.0_DYToll01234Jets_5f_LO_MLM_Madgraph_LHE_13TeV+DYToll01234Jets_5f_LO_MLM_Madgraph_LHE_13TeV+Hadronizer_TuneCP5_13TeV_MLM_5f_max4j_LHE_pythia8+NANOGENFromGen
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-84e32b/548.0_TTbar_Pow_LHE_13TeV+TTbar_Pow_LHE_13TeV+Hadronizer_TuneCP5_13TeV_powhegEmissionVeto2p_pythia8+NANOGENFromGen

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 42
 - DQMHistoTests: Total histograms compared: 3250704
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3250682
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 41 files compared)
 - Checked 177 log files, 37 edm output root files, 42 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2021-12-09-1100+84e32b/47370/triggerResults): no differences found

mariadalfonso commented:
+xpog

* added gen Filter Efficiency in the lumi branch
https://github.com/cms-sw/cmssw/pull/34169#issuecomment-988063718
* size increase negligible
* central nano and nanoGen updated

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1
- After a long digestion, the code got finally blessed and approved by xpog.
- There are however several unanswered relevant remarks, that should be addressed. Normally "it was as such in the original code from which I derived my development" cannot be accepted as an excuse to avoid code and performance improvements and fixes. However, given the longish story of this PR, we can accept to merge it as such and allow it stably in the release.
- A github issue with the remaining points to be addressed was opened in #36461. Either @mzarucki or any other developer designed by @cms-sw/xpog-l2 is expected to take care and implement them at the earliest
